// @plasmicapp/react-web/skinny/dist/ssr-9ebd98ed.js
import { SSRProvider, useIsSSR as useIsSSR$1 } from '@react-aria/ssr';
import * as React from 'react';
import React__default from 'react';

var PlasmicHeadContext = React.createContext(undefined);
function PlasmicHead(props) {
    var Head = React.useContext(PlasmicHeadContext);
    if (!Head) {
        console.warn("Plasmic: Head meta tags are being ignored. To make them work, pass a Head component into PlasmicRootProvider.");
        // TODO: Link to doc about Head.
        return null;
    }
    return (React.createElement(Head, null,
        props.image ? (React.createElement(React.Fragment, null,
            React.createElement("meta", { name: "twitter:card", content: "summary_large_image" }),
            React.createElement("meta", { key: "og:image", property: "og:image", content: props.image }),
            React.createElement("meta", { key: "twitter:image", name: "twitter:image", content: props.image }))) : (React.createElement("meta", { name: "twitter:card", content: "summary" })),
        props.title && (React.createElement(React.Fragment, null,
            React.createElement("title", { key: "title" }, props.title),
            React.createElement("meta", { key: "og:title", property: "og:title", content: props.title }),
            React.createElement("meta", { key: "twitter:title", property: "twitter:title", content: props.title }))),
        props.description && (React.createElement(React.Fragment, null,
            React.createElement("meta", { key: "description", name: "description", content: props.description }),
            React.createElement("meta", { key: "og:description", property: "og:description", content: props.description }),
            React.createElement("meta", { key: "twitter:description", name: "twitter:description", content: props.description }))),
        props.canonical && React.createElement("link", { ref: "canonical", href: props.canonical })));
}
var plasmicHeadMeta = {
    name: "hostless-plasmic-head",
    displayName: "Head",
    description: "Used to add page metadata to HEAD tag",
    importName: "PlasmicHead",
    importPath: "@plasmicapp/react-web",
    props: {
        title: {
            type: "string",
            displayName: "Title",
        },
        description: {
            type: "string",
            displayName: "Description",
        },
        image: {
            type: "imageUrl",
            displayName: "Image",
        },
        canonical: {
            type: "string",
            displayName: "Canonical URL",
        },
    },
};

var PlasmicTranslatorContext = React__default.createContext(undefined);
function isIterable(val) {
    return val != null && typeof val[Symbol.iterator] === "function";
}
function genTranslatableString(elt) {
    var components = {};
    var componentsCount = 0;
    var getText = function (node) {
        if (!node) {
            return "";
        }
        if (typeof node === "number" ||
            typeof node === "boolean" ||
            typeof node === "string") {
            return node.toString();
        }
        if (typeof node !== "object") {
            return "";
        }
        if (Array.isArray(node) || isIterable(node)) {
            return Array.from(node)
                .map(function (child) { return getText(child); })
                .filter(function (child) { return !!child; })
                .join("");
        }
        var nodeChildren = (hasKey(node, "props") &&
            hasKey(node.props, "children") &&
            node.props.children) ||
            (hasKey(node, "children") && node.children) ||
            [];
        var contents = "" + React__default.Children.toArray(nodeChildren)
            .map(function (child) { return getText(child); })
            .filter(function (child) { return !!child; })
            .join("");
        if (React__default.isValidElement(node) && node.type === React__default.Fragment) {
            return contents;
        }
        var componentId = componentsCount + 1;
        componentsCount++;
        components[componentId] = React__default.isValidElement(node)
            ? React__default.cloneElement(node, {
                key: componentId,
                children: undefined,
            })
            : node;
        return "<" + componentId + ">" + contents + "</" + componentId + ">";
    };
    var str = getText(elt);
    return {
        str: str,
        components: components,
        componentsCount: componentsCount,
    };
}
function Trans(_a) {
    var children = _a.children;
    var _t = React__default.useContext(PlasmicTranslatorContext);
    if (!_t) {
        warnNoTranslationFunctionAtMostOnce();
        return children;
    }
    var _b = genTranslatableString(children), str = _b.str, components = _b.components, componentsCount = _b.componentsCount;
    return _t(str, componentsCount > 0 ? { components: components } : undefined);
}
var hasWarned = false;
function warnNoTranslationFunctionAtMostOnce() {
    if (!hasWarned) {
        console.warn("Using Plasmic Translation but no translation function has been provided");
        hasWarned = true;
    }
}
function hasKey(v, key) {
    return typeof v === "object" && v !== null && key in v;
}

var PlasmicRootContext = React.createContext(undefined);
function PlasmicRootProvider(props) {
    var platform = props.platform, children = props.children;
    var context = React.useMemo(function () { return ({
        platform: platform,
    }); }, [platform]);
    return (React.createElement(PlasmicRootContext.Provider, { value: context },
        React.createElement(SSRProvider, null,
            React.createElement(PlasmicTranslatorContext.Provider, { value: props.translator },
                React.createElement(PlasmicHeadContext.Provider, { value: props.Head }, children)))));
}
var useIsSSR = useIsSSR$1;
function useHasPlasmicRoot() {
    return !!React.useContext(PlasmicRootContext);
}
var hasWarnedSSR = false;
/**
 * Warns the user if PlasmicRootProvider is not used
 */
function useEnsureSSRProvider() {
    var hasRoot = useHasPlasmicRoot();
    if (hasRoot || hasWarnedSSR || process.env.NODE_ENV !== "development") {
        return;
    }
    hasWarnedSSR = true;
    console.warn("Plasmic: To ensure your components work correctly with server-side rendering, please use PlasmicRootProvider at the root of your application.  See https://docs.plasmic.app/learn/ssr");
}

export { PlasmicHead as P, Trans as T, PlasmicRootProvider as a, useEnsureSSRProvider as b, genTranslatableString as g, plasmicHeadMeta as p, useIsSSR as u };
